core/src/test/scala/unit/kafka/network/SocketServerTest.scala

```diff
@@ -179,15 +179,20 @@ class SocketServerTest extends JUnitSuite {
 
   @Test
   def testSocketsCloseOnShutdown() {
+    import java.util.concurrent.TimeUnit
     // open a connection
     val plainSocket = connect(protocol = SecurityProtocol.PLAINTEXT)
+    plainSocket.setTcpNoDelay(true)
     val traceSocket = connect(protocol = SecurityProtocol.TRACE)
+    traceSocket.setTcpNoDelay(true)
     val bytes = new Array[Byte](40)
     // send a request first to make sure the connection has been picked up by the socket server
     sendRequest(plainSocket, bytes, Some(0))
     sendRequest(traceSocket, bytes, Some(0))
     processRequest(server.requestChannel)
-
+    // Sleep for 200ms to not run into TCP SLOW_START behaviour potentially masking issues in
+    // sending data below
+    TimeUnit.MILLISECONDS.sleep(200L)
```

Comment:
@ijuma no `-1` as I expected. Tried:

```scala
  def testSocketsCloseOnShutdown() {
    // open a connection
    val plainSocket = connect(protocol = SecurityProtocol.PLAINTEXT)
    plainSocket.setTcpNoDelay(true)
    val traceSocket = connect(protocol = SecurityProtocol.TRACE)
    traceSocket.setTcpNoDelay(true)
    val bytes = new Array[Byte](40)
    // send a request first to make sure the connection has been picked up by the socket server
    sendRequest(plainSocket, bytes, Some(0))
    sendRequest(traceSocket, bytes, Some(0))
    processRequest(server.requestChannel)
    // make sure the sockets are open
    server.acceptors.values.map(acceptor => assertFalse(acceptor.serverChannel.socket.isClosed))
    // then shutdown the server
    server.shutdown()
    while(plainSocket.getInputStream.read() > -1) {
    }
```

And run into:

```sh
java.net.SocketException: Connection reset

	at java.net.SocketInputStream.read(SocketInputStream.java:210)
	at java.net.SocketInputStream.read(SocketInputStream.java:141)
	at java.net.SocketInputStream.read(SocketInputStream.java:224)
	at kafka.network.SocketServerTest.testSocketsCloseOnShutdown(SocketServerTest.scala:196)
```

Also tried to suppress here:

```scala
    try {
      while (plainSocket.getInputStream.read() > -1) {}
    } catch {
      case _: IOException => // expected
    }
```

But this starts blocking forever on the read eventually as seen below:

![blocks](https://cloud.githubusercontent.com/assets/6490959/23578121/e7c2ac4a-00cf-11e7-9de7-55f80187772f.png)

Also even if we handle this with timeouts like so (would have to do this to both sockets, it doesn't work reliably either
![both](https://cloud.githubusercontent.com/assets/6490959/23578137/65c9cdd0-00d0-11e7-938b-b65a8c8b4d09.png)

I think the explanation for that can be found here (shutdownOutput != connection shut down):
http://stackoverflow.com/a/6990656/5509152

-> I think I like no nagle + pessimistically sized 200ms wait still tbh.